app/src/main/java/pl/lambada/songsync/data/remote/lyrics_providers/LyricsProviderService.kt

```diff
@@ -0,0 +1,104 @@
+import pl.lambada.songsync.util.EmptyQueryException
+import pl.lambada.songsync.util.InternalErrorException
+import pl.lambada.songsync.util.NoTrackFoundException
+import pl.lambada.songsync.util.Providers
+import java.io.FileNotFoundException
+import java.net.UnknownHostException
+
+/**
+ * Service class for interacting with different lyrics providers.
+ */
+class LyricsProviderService {
+    // Spotify API token
+    private val spotifyAPI = SpotifyAPI()
+
+    // LRCLib Track ID
+    private var lrcLibID = 0
+
+    // Netease Track ID and stuff
+    private var neteaseID = 0L
+
+    // Apple Track ID
+    private var appleID = 0L
+    // TODO: Use values from SongInfo object returned by search instead of storing them here
+
+    /**
+     * Refreshes the access token by sending a request to the Spotify API.
+     */
+    suspend fun refreshSpotifyToken() = kotlin.runCatching {
+        spotifyAPI.refreshToken()
+    }
+
+    /**
+     * Gets song information from the Spotify API.
+     * @param query The SongInfo object with songName and artistName fields filled.
+     * @param offset (optional) The offset used for trying to find a better match or searching again.
+     * @return The SongInfo object containing the song information.
+     */
+    @Throws(
+        UnknownHostException::class, FileNotFoundException::class, NoTrackFoundException::class,
+        EmptyQueryException::class, InternalErrorException::class
+    )
+    suspend fun getSongInfo(query: SongInfo, offset: Int? = 0, provider: Providers): SongInfo? {
+        return try {
+            when (provider) {
+                Providers.SPOTIFY -> spotifyAPI.getSongInfo(query, offset)
+                Providers.LRCLIB -> LRCLibAPI().getSongInfo(query).also {
+                    lrcLibID = it?.lrcLibID ?: 0
+                } ?: throw NoTrackFoundException()
+
+                Providers.NETEASE -> NeteaseAPI().getSongInfo(query, offset).also {
+                    neteaseID = it?.neteaseID ?: 0
+                } ?: throw NoTrackFoundException()
+
+                Providers.APPLE -> AppleAPI().getSongInfo(query).also {
+                    appleID = it?.appleID ?: 0
+                } ?: throw NoTrackFoundException()
+            }
+        } catch (e: InternalErrorException) {
+            throw e
+        } catch (e: NoTrackFoundException) {
+            throw e
+        } catch (e: EmptyQueryException) {
+            throw e
+        } catch (e: Exception) {
+            throw InternalErrorException(Log.getStackTraceString(e))
+        }
+    }
+
+    /**
+     * Gets synced lyrics using the song link and returns them as a string formatted as an LRC file.
+     * @param songLink The link to the song.
+     * @return The synced lyrics as a string.
+     */
+    suspend fun getSyncedLyrics(
+        songLink: String,
+        version: String,
+        provider: Providers,
+        // TODO providers could be a sealed interface to include such parameters
```

Comment:
True, but let's leave this to another PR

Reply:
feel free to merge if you dont find any other issues